adapters/Unity/UnityAdapter/GADMUnityInterstitialAd.m

```diff
@@ -0,0 +1,194 @@
+  __weak id<GADMAdNetworkConnector> _connector;
+  
+  /// Adapter for receiving ad request notifications.
+  __weak id<GADMAdNetworkAdapter> _adapter;
+}
+
+  /// A dictionary to keep track loaded Placement IDs
+  static NSMutableDictionary<NSString *, id> *_placementInUse;
+
+- (instancetype)initWithGADMAdNetworkConnector:(nonnull id<GADMAdNetworkConnector>)connector
+                                       adapter:(nonnull id<GADMAdNetworkAdapter>)adapter {
+  self = [super init];
+  if (self) {
+    _adapter = adapter;
+    _connector = connector;
+    if (_placementInUse == nil) {
+      _placementInUse = [[NSMutableDictionary alloc] init];
+    }
+  }
+  return self;
+}
+
+- (void)getInterstitial {
+  id<GADMAdNetworkConnector> strongConnector = _connector;
+  id<GADMAdNetworkAdapter> strongAdapter = _adapter;
+  
+  if (!strongConnector) {
+    NSLog(@"Unity Ads Adapter Error: No GADMAdNetworkConnector found.");
+    return;
+  }
+  
+  if (!strongAdapter) {
+    NSLog(@"Unity Ads Adapter Error: No GADMAdNetworkAdapter found.");
+    return;
+  }
+  
+  _gameID = [[[strongConnector credentials] objectForKey:kGADMAdapterUnityGameID] copy];
+  _placementID = [[[strongConnector credentials] objectForKey:kGADMAdapterUnityPlacementID] copy];
+  if (!_gameID || !_placementID) {
+     NSError *error = GADMAdapterUnityErrorWithCodeAndDescription(
+           GADMAdapterUnityErrorInvalidServerParameters, @"Game ID and Placement ID cannot be nil.");
+    [strongConnector adapter:strongAdapter didFailAd:error];
+    return;
+  }
+
+  if (![UnityAds isInitialized]) {
+    [[GADMAdapterUnity alloc] initializeWithGameID: _gameID withInitDelegate:Nil];
+  }
+
+  @synchronized (_placementInUse) {
+    if ([_placementInUse objectForKey:_placementID]) {
+      if (strongConnector && strongAdapter) {
+        NSString *errorMsg = [NSString stringWithFormat:@"An ad is already loading for placement ID: %@.", _placementID];
+        NSError *error = GADMAdapterUnityErrorWithCodeAndDescription(GADMAdapterUnityErrorAdAlreadyLoaded, errorMsg);
+        [strongConnector adapter:strongAdapter didFailAd:error];
+      }
+      return;
+    }
+
+    [_placementInUse setValue:self forKey:_placementID];
+}
+  [UnityAds load:_placementID loadDelegate:self];
+}
+
+- (void)presentInterstitialFromRootViewController:(UIViewController *)rootViewController {
+  @synchronized (_placementInUse) {
+    [_placementInUse removeObjectForKey:_placementID];
+  }
+  // We will send adapterWillPresentInterstitial callback before presenting unity ad because the ad has already loaded.
+  id<GADMAdNetworkConnector> strongConnector = _connector;
+  id<GADMAdNetworkAdapter> strongAdapter = _adapter;
+  
+  if (!strongConnector) {
+    NSLog(@"Unity Ads Adapter Error: No GADMAdNetworkConnector found.");
+    return;
+  }
+  
+  if (!strongAdapter) {
+    NSLog(@"Unity Ads Adapter Error: No GADMAdNetworkAdapter found.");
+    return;
+  }
+  
+  if ([UnityAds isReady:_placementID]) {
```

Comment:
if (![UnityAds isReady:_placementID]) {
    NSError *error = GADMAdapterUnityErrorWithCodeAndDescription(GADMAdapterUnityErrorShowAdNotReady, @"Failed to show Unity Ads interstitial.");
    [_adEventDelegate didFailToPresentWithError:error];
    return;
  }